repo/blob/s3/s3_storage.go

```diff
@@ -312,6 +312,42 @@ func (s *s3Storage) FlushCaches(ctx context.Context) error {
 	return nil
 }
 
+func isAccessDenied(err error) bool {
+	var e minio.ErrorResponse
+	return errors.As(err, &e) && e.Code == "AccessDenied"
+}
+
+func (s *s3Storage) bucketExists(ctx context.Context) (bool, error) {
+	ok, err := s.cli.BucketExists(ctx, s.Options.BucketName)
+	if err == nil {
+		return ok, nil
+	}
+
+	if !isAccessDenied(err) {
+		return false, errors.Wrap(err, "error checking bucket existence")
+	}
+
+	if s.Options.Prefix == "" {
```

Comment:
Can the check below (for a non-existing blob) be performed as well when the prefix is empty?

Reply:
When the execution comes to this `if` block (`if s.Options.Prefix == ""`), it means that `s.cli.BucketExists` has returned an `AccessDenied` error. This already indicates that there is no permission to access the bucket itself. 
Then if the prefix is empty, it means users want to write data to the bucket directly (not to a sub dir), however, since we've proved that directly accessing to the bucket is impossible, it is safe to return an error here.

Reply:
For this function (`checkBucketAccess`), the purpose still abides the original one --- check the existence of the bucket. Merely, because of the exception described in #3071, there is a false report if we only call `s.cli.BucketExists`. 
We read against the prefix, is not to assure the Get permission, but to assure the existence of the bucket dir -- `<bucket name>/<prefix>`. Basically, we can use any operation to do the check, i.e., `PutBlob`, `ListBlobs`, while get a non-existing blob is with the least cost.

Or in another word, this function is an EX version of the original `s.cli.BucketExists` function, which assures the existence of the bucket and works in all cases.

We don't aim to check all the permissions that will be used by the later repo operations. If there is indeed a permission error for further repo operations, the error will be returned along with the operation itself, e.g., following this function, we will call `GetBlob` to get the `.storageconfig` file, where the read object permission will be checked.

Reply:
Let me know if we want to do more checks based on the original purpose.

Reply:
The request is to either:
- Preferred: not perform the blob existence check at all when `BucketExists(...)` returns access denied. Simply log, and `return nil`. See below.
- Or perform the blob existence check in all cases, not just when the prefix is empty. This is less preferred, because of the reasons explained in the comment below.


```Go
	if isAccessDenied(err) {
		// log.Infof("unable to determine if bucket %q exists, error: %s", s.Options.BucketName, err)
		return nil
	}
```